include/Target/Btor/BtorToBtorIRTranslation.h

```diff
@@ -0,0 +1,291 @@
+      if (m_modelFile) {
+        fclose(m_modelFile);
+      }
+  }
+
+///===----------------------------------------------------------------------===//
+/// Parse btor2 file
+///===----------------------------------------------------------------------===//
+  
+  bool parseModelIsSuccessful();
+
+///===----------------------------------------------------------------------===//
+/// Create MLIR module
+///===----------------------------------------------------------------------===//
+  
+  Value getFromCacheById(const int64_t id) {
+    assert(valueAtIdIsInCache(id));
+    return m_cache.at(id);
+  }
+
+  void setCacheWithId(const int64_t id, const Value &value) {
+    assert(!valueAtIdIsInCache(id));
+    m_cache[id] = value;
+    assert(valueAtIdIsInCache(id));
+  }
+
+  void setCacheWithId(const int64_t id, Operation * op) {
+    // We never have to use the result of a btor line with no 
+    // return values since btor2 doesn't allow it
+    if (hasReturnValue(getLineById(id))) {
+      assert(op);
+      assert(op->getNumResults() == 1);
+      assert(op->getResult(0));
+      setCacheWithId(id, op->getResult(0));
+    }
+  }
+
+  bool valueAtIdIsInCache(const int64_t id) {
+    return m_cache.count(id) != 0;
+  }
+
+  OwningOpRef<FuncOp> buildInitFunction();
+  OwningOpRef<FuncOp> buildNextFunction();
+  
+ private: 
+///===----------------------------------------------------------------------===//
+/// Parse btor2 file
+///===----------------------------------------------------------------------===//
+  
+  Btor2Parser *m_model = nullptr;
+  FILE *m_modelFile = nullptr;
+
+  std::vector<Btor2Line *> m_states;
+  std::vector<Btor2Line *> m_bads;
+  std::vector<Btor2Line *> m_inits;
+  std::vector<Btor2Line *> m_nexts;
+  std::vector<Btor2Line *> m_constraints;
+  std::vector<Btor2Line *> m_lines;
+
+  std::map<int64_t, Value> m_cache;
+
+  void parseModelLine(Btor2Line *l);
+
+  Btor2Line * getLineById(unsigned id) {
+      assert(id < m_lines.size());
+      return m_lines.at(id);
+  }
+
+  void setLineWithId(unsigned id, Btor2Line * line) {
+      assert(id < m_lines.size());
+      assert(!m_lines.at(id));
+      m_lines[id] = line;
+  }
+///===----------------------------------------------------------------------===//
+/// Create MLIR module
+///===----------------------------------------------------------------------===//
+
+  MLIRContext *m_context;
+  OpBuilder m_builder;
+  Location m_unknownLoc;
+  
+  void toOp(Btor2Line *line);
+  bool needsMLIROp(Btor2Line * line);
+  bool hasReturnValue(Btor2Line * line);
+  void createNegateLine(int64_t curAt, const Value &child);
+  Operation * createMLIR(const Btor2Line *line, 
+                        const SmallVector<Value> &kids,
+                        const SmallVector<unsigned> &arguments);
+
+  // Builder wrappers
+  Type getIntegerTypeOf(Btor2Line * line) {
+    return m_builder.getIntegerType(line->sort.bitvec.width);
+  }
+
+  std::vector<Value> collectReturnValuesForInit(
```

Comment:
same question as before, are mlir types to be passed by value or reference. If by reference, then vectors should be `std::vector<Value*>` and `std::vector<Type*>`